content/posts/aws-autoscaling-groups-with-terraform.md

```diff
@@ -0,0 +1,193 @@
+---
+title: "Updating AWS autoscaling groups with Terraform and preserve desired capacity"
+date: 2020-06-21
+authors:
+  - author:
+      name: "Ariel Peltz"
+      link: "https://github.com/arielpeltz"
+tags: ["aws", "terraform", "autoscaling", "desired_capacity"]
+---
+
+## Introduction
+Terraform is probably the best tool for deploying infrastructure.  So it was an obvious choice
+when we wanted to deploy an autoscaling group into AWS.  The project we are working on requires us to
+deploy a service on instances in AWS.  The service is stateless and has simple configuration that is
+easy to configure using [cloud-init](https://cloudinit.readthedocs.io/en/latest/index.html).
+
+Our intention is to be able to configure everything with Terraform. Every time we need to make
+a configuration change, we update Terraform configuration and apply again.
+We wanted to have clean instances when we make any configuration change and not update in place.
+
+> Throughout the examples we omitted some configuration elements creation for brevity.
+> Others are injected as variables and are also not shown here.
+
+## Initial Configuration
+First create the launch template
+```terraform
+# Query for the latest Ubuntu 18.04 AMI in the region
+data "aws_ami" "ubuntu" {
+  most_recent = true
+
+  filter {
+    name   = "name"
+    values = ["ubuntu/images/hvm-ssd/ubuntu-bionic-18.04-amd64-server-*"]
+  }
+
+  filter {
+    name   = "virtualization-type"
+    values = ["hvm"]
+  }
+
+  owners = ["099720109477"] # Canonical
+}
+
+resource "aws_launch_template" "mysvc" {
+  name                                 = "mysvc"
+  ebs_optimized                        = true
+  image_id                             = data.aws_ami.ubuntu.id
+  instance_initiated_shutdown_behavior = "terminate"
+  instance_type                        = "c5.large"
+  key_name                             = var.key_name
+  network_interfaces {
+    associate_public_ip_address = true
+    subnet_id                   = var.subnet_id
+    security_groups             = var.security_groups
+    delete_on_termination       = true
+  }
+  tag_specifications {
+    resource_type = "instance"
+    tags = {
+      Name          = "mysvc"
+    }
+  }
+  user_data = base64encode(templatefile("${path.module}/user_data.yaml", {
+    web_password    = var.web_password,
+    ssl_private_key = tls_private_key.mysvc.private_key_pem,
+    ssl_certificate = tls_self_signed_cert.mysvc.cert_pem
+  }))
+
+  # we don't want to create a new template just because there is a newer AMI
+  lifecycle {
+    ignore_changes = [
+      image_id,
+    ]
+  }
+}
+
+```
+The launch template defines the critical elements of the instance but also loads
```

Comment:
Why not "and also" instead of "but also"? Is there any contrast between the two?

Reply:
You are right, fixed